File: Algoexpert/Reverse Linked List/cpp/reverse_linked_list.cpp

#include <bits/stdc++.h>

using namespace std;

class LinkedList {
    public:
    int value;
    LinkedList *next;

    LinkedList(int value) {
        this->value = value;
        this->next = nullptr;
    }
};

LinkedList *reverseLinkedList(LinkedList *head) {
    LinkedList *previousNode = nullptr;
    LinkedList *currentNode = head;
    while(currentNode!=nullptr) {
        LinkedList *nextNode = currentNode->next;
        currentNode->next = previousNode;
        previousNode = currentNode;
        currentNode = nextNode;
    }
    return previousNode;
}

LinkedList *newLinkedList(vector<int> values);
vector<int> toArray(LinkedList *ll);
bool arraysEqual(vector<int> arr1, vector<int> arr2);

LinkedList *newLinkedList(vector<int> values) {
    LinkedList *ll = new LinkedList(values[0]);
    LinkedList *current = ll;
    for (int i = 1; i < values.size(); i++) {
        current->next = new LinkedList(values[i]);
        current = current->next;
    }
    return ll;
}

vector<int> toArray(LinkedList *ll) {
    vector<int> arr = {};
    LinkedList *current = ll;
    while (current != nullptr) {
        arr.push_back(current->value);
        current = current->next;
    }
    return arr;
}

bool arraysEqual(vector<int> arr1, vector<int> arr2) {
    if (arr1.size() != arr2.size())
        return false;
    for (int i = 0; i < arr1.size(); i++) {
        if (arr1[i] != arr2[i])
        return false;
    }
    return true;
}

int main() {

    LinkedList *test = newLinkedList({0, 1, 2, 3, 4, 5});
    vector<int> result = toArray(reverseLinkedList(test));
    vector<int> expected = {5, 4, 3, 2, 1, 0};
    assert(arraysEqual(result, expected));

    return 0;
}

File: Algoexpert/Reverse Linked List/java/reverseLinkedList.java
public class reverseLinkedList {

    public static LinkedList reverseLinkedList(LinkedList head) {

        LinkedList previousNode = null;
        LinkedList currentNode = head;
        while (currentNode != null) {
            LinkedList nextNode = currentNode.next;
            currentNode.next = previousNode;
            previousNode = currentNode;
            currentNode = nextNode;
        }
        return previousNode;
    }

    public static void main(String[] args) {

        LinkedList node1 = new LinkedList(1);
        LinkedList node2 = new LinkedList(2);
        LinkedList node3 = new LinkedList(3);
        LinkedList node4 = new LinkedList(4);

        node1.next = node2;
        node2.next = node3;
        node3.next = node4;

        LinkedList head = reverseLinkedList(node1);

        while (head != null) {
            System.out.println(head.val);
            head = head.next;
        }

    }

}
